Adding Metadata Operations sample

JimDaly opened the pull request:
This sample demonstrates how to perform a wide range of data operations related to metadata and solutions.

kalavany commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2312 @@
+        /// Users are always guaranteed to read the latest committed write.
+        /// Use of this header incurs a small overhead and is therefore not suited to high volume workloads.
+        /// </summary>
+        private static readonly Dictionary<string, List<string>> STRONGCONSISTENCYHEADER = new Dictionary<string, List<string>>()
```
Can we name it strongConsistencyHeader?

 private static readonly Dictionary<string, List<string>> strongConsistencyHeader = new Dictionary<string, List<string>>()

JimDaly replied:
Done

kalavany commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2312 @@
+            Guid solutionComponentId;
+
+            if (!string.IsNullOrEmpty(solutionUniqueName))
+            {
```
It's recommended to throw the exception in if clause. Then there would be no else clause.

if (string.IsNullOrEmpty(solutionUniqueName))
            {
throw new Exception("AddSolutionComponent requires the solutionUniqueName parameter value.");
}

JimDaly replied:
Done

phecke commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2458 @@
+            }
+            return solutionComponentId;
+        }
+
```
The next 3 methods can each be reduced to one call to svc.Post() with a inline anonymous JObject definition without the need to code a method. This is more a matter of the author's writing style and is just mentioned here as an FYI.

JimDaly replied:
Thanks. 
While each of these have a similar pattern, the goal is to show how to use the named actions rather than write a minimal amount of code.
[CanBeReferenced Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canbereferenced?view=dynamics-ce-odata-9)
[CanBeReferencing Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canbereferencing?view=dynamics-ce-odata-9)
[CanManyToMany Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canmanytomany?view=dynamics-ce-odata-9)

phecke commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2458 @@
+                {
+                    solutionId = svc.CheckSolution(solutionUniqueName);
+                }
+                catch (Exception)
```
Why catch an exception and do nothing more than throw it? Perhaps this was a placeholder for you to add some code. Anyway, this shows up in many places in your code. I would recommend removing it unless you are doing more than throwing the caught exception.

JimDaly replied:
This code is calling another method `CheckSolution ` which throws a detailed exception.
This is just verifying that the solution they named is valid before proceeding. 
Yes, it is called this way multiple times because it provides the same solution checking each time the solutionUniqueName parameter is passed.

phecke commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2458 @@
+            List<ComplexOneToManyRelationshipMetadata> oneToManyRelationships,
+            string solutionUniqueName = null)
+        {
+            if (!string.IsNullOrEmpty(solutionUniqueName))
```
You are checking if the solution name is null or empty several times in this method. Best to check once at the beginning of the method, and also call CheckSolution(). Once that is done you are good to go for the rest of the method.

JimDaly replied:
This operation is to create a customer relationship and solutionUniqueName is an optional parameter which, if included, will associate the 3 created solution components with the unmanaged solution indicated by the solutionUniqueName.

- First check is to ensure that the solution exists and is managed.
- Second check is to only include the value in the JObject body when it exists
- Final check is after the 3 solution components are created, then they can be associated with the named solution if it exists.

In other words, the association with a solution is a secondary, optional operation within the flow. It doesn't control the main flow, just supplements it when provided. Therefore, it seems appropriate to use the existence of the parameter when needed (3 times) to control the secondary optimal options.

phecke commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2458 @@
+        public static void DeleteEntity(this CDSWebApiService svc,
+            Guid entityMetadataId)
+        {
+            svc.DeleteEntity(entityMetadataId, null);
```
For this method and the ones that follow, it is not recommended to have methods that do nothing other than call another method.

JimDaly replied:
There are two mutually exclusive ways to access many metadata items: by Metadata Id or by Name which is an alternate key.
See: [Retrieve metadata by name or MetadataId](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/retrieve-metadata-name-metadataid)

These methods do more than just call another method, they provide overrides that can be easily used depending on the developer's situation. One accepts a Guid id value, the other a string name value. Both of them call the same private method that can accept both and manage passing the correct one.

It is intentional and a means to simplify the developer experience while re-using the same underlying logic.

phecke commented on cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:
```diff
@@ -0,0 +1,2458 @@
+                    ["EntityName"] = entityName
+                };
+                return (bool)svc.Post("CanBeReferenced", body)["CanBeReferenced"];
+            }
```
General observation that is replicated throughout this code file. You have try/catch blocks in which you create a new exception with the caught exception as the inner exception and the parent exception message contains the inner exception message. Or you create an exception that tells the end user what method the exception occurred in. Neither of these are needed as you will be able to find this info when using a debugger.

Perhaps you like having such info printed out in the console without having to use a debugger? The problem with this approach is that 1) our dev customers know and expect to use a debugger, and 2) if they want to copy and use the code in their own projects (which they do) they will have to strip this code out.

JimDaly replied:
I'd be happy to discuss alternative patterns. The error handling I've done is focused on identifying the source of the error while debugging the console application sample. This is somewhere in-between the raw-low level samples and an SDK.

If you have a different pattern that you would recommend, please submit a PR and I can review it.

phecke reviewed (changes requested):
I went through the first 1000 lines in the Extensions.cs file (see inline comments), plus my comments for the CDSWebApiService.cs file in email. That is all I had the time for right now.

JimDaly commented:
@phecke 
Do you have any specific changes you want made?
I think I've addressed your comments and I responded to your questions in email.
I don't see anything that has to block this PR so if I don't hear back from you in a couple days I will go ahead and merge.

phecke commented:
@JimDaly My comments were suggestions. No significant problems identified. Go ahead and merge the PR.
